### DISCUSSION TITLE
Feature Request: Custom terms of Service Agreements before Invoice is presented.

### DISCUSSION THREAD

**Discussion by beeforbacon** [btcpayserver/btcpayserver#5002] [Ideas / Feature Requests]
I run a business where the customer must agree to a waiver before paying. I would like the option on BTCPayServer to have a required ToS checkbox, which is configured in the BTCPayServer Apps page. Similar to all the other customized features. After payment is made, in the invoices menu, there is a field that says that customer checked agreed on the ToS. This would really help and hopefully isn't too much dev time required. Let me know if anyone needs further information or clarification. 

Example pics of what I am looking for below: 

![image](https://github.com/btcpayserver/btcpayserver/assets/51864803/64f39069-6b9c-4d46-9ff8-3135842c772b)

![image](https://github.com/btcpayserver/btcpayserver/assets/51864803/77868f3a-1c9e-40cb-a541-c5012181412d)

**Comment by mp21-org** [comment]
Sounds like a useful feature. Tried doing this with the form builder and could only create a select box (with dropdown list options "ACK" and "NACK") and also no chance of getting a link in there. I assume you are using the PoS App and not i.e. Woocommerce or such.

We'll discuss and let you know if and when it will come. But persumably not in 1.10.0 which is right around the corner and in feature freeze.

**Reply by beeforbacon** [reply]
For the form customization, for ACK/NACK response, if NACK can the page just go back to the app store page instead of continuing? If so, how would the json look? All I want is "Do you agree to terms of service from <website>" then drop down yes or no, if no, then dont allow to continue. That would work, I just don't know how to do the json format for that. I just want "yes" to be the only acceptable reply to be taken to the invoice and then recorded in the invoices page.

**Reply by mp21-org** [reply]
The fact that the html is stripped from the label but not from the error text is probably a bug ^^
![grafik](https://github.com/btcpayserver/btcpayserver/assets/1374810/47acf012-0e60-4ebc-8b74-524add9fa656)

```json
{
  "topMessages": [],
  "fields": [
    {
      "name": "newField1",
      "constant": false,
      "type": "select",
      "value": "newOption2",
      "required": true,
      "label": "Agree to our <a href=\"tos.html\">TOS</a>",
      "originalValue": null,
      "helpText": null,
      "validationErrors": [],
      "fields": [],
      "options": [
        {
          "text": "NACK",
          "value": ""
        },
        {
          "text": "ACK",
          "value": "ack"
        }
      ]
    }
  ]
}
```

**Reply by beeforbacon** [reply]
The json you presented does not show any text enter fields for me. Do I need to edit something?

**Reply by mp21-org** [reply]
That was just regarding the TOS accept ability. You would add other fields you need accordingly.
There will be a visual form editor in the next release. sneak peek: #4968

**Reply by beeforbacon** [reply]
Alright thanks. My issue with the form creator right now is that no matter what is entered, the invoice is presented. I want only when "yes" is entered does the invoice present itself. I just can't figure it out. It would be a fine hold-me-over until the day the ToS checkbox is implemented :)

**Reply by mp21-org** [reply]
strange. just tested with 1.9.3 and I only get so an invoice when I excplicitly select "ACK":
https://pay.4sats.net/forms/6f6b9ce8-9366-41e1-aa46-b7c57f62e8ef

**Reply by beeforbacon** [reply]
looks like im not fully updated...will update now.

**Reply by beeforbacon** [reply]
Okay it's working... my bad on that. Thanks for the help petzsch, it means a lot!
